### PR TITLE
Don't upload replays for failed scores

### PR DESCRIPTION
**Pull request by smoogipoo**
Resolves https://github.com/ppy/osu-server-spectator/issues/187
Closes https://github.com/ppy/osu-infrastructure/issues/19

`SoloScore` model taken from [osu-queue-score-statistics](https://github.com/ppy/osu-queue-score-statistics/blob/3cc58af62a4805ca51c1ef13ed85385b4179cf84/osu.Server.Queues.ScoreStatisticsProcessor/Models/SoloScore.cs)

### CODE REVIEW

File: osu.Server.Spectator/Database/DatabaseAccess.cs [hunk 1]
```diff
@@ -295,14 +295,15 @@ public async Task MarkScoreHasReplay(Score score)
...
 
-            return await connection.QuerySingleOrDefaultAsync<long?>("SELECT `score_id` FROM `solo_score_tokens` WHERE `id` = @Id", new
-            {
-                Id = token
-            });
+            return await connection.QuerySingleOrDefaultAsync<SoloScore?>(
+                "SELECT * FROM `solo_scores` WHERE `id` = (SELECT `score_id` FROM `solo_score_tokens` WHERE `id` = @Id)", new
+                {
+                    Id = token
+                });
```
Comment (bdach) [inline osu.Server.Spectator/Database/DatabaseAccess.cs:306]: I'll just note that this is changing the same method that my series of pulls concerning replay treatment is changing (see https://github.com/ppy/osu-server-spectator/pull/185).

This query is still - same as `master` - correct only for solo scores and wrong for multiplayer scores (see https://github.com/ppy/osu/pull/24697 / https://github.com/ppy/osu-infrastructure/blob/master/score-submission.md#multiplayer for full explanation), but I suppose it is no *more* wrong than `master`.

Make of that what you will wrt which pull should go first etc.

File: osu.Server.Spectator/Hubs/ScoreUploader.cs [hunk 2]
```diff
@@ -101,14 +102,26 @@ public async Task Flush()
...
+                                return;
+                            }
+
+                            if (!dbScore.ScoreInfo.Passed)
+                                return;
+
+                            var updatedScore = new Score
+                            {
+                                Replay = item.Score.Replay,
+                                ScoreInfo = dbScore.ScoreInfo.ToScoreInfo(item.Score.ScoreInfo.Mods, item.Score.ScoreInfo.BeatmapInfo)
```
Comment (bdach) [inline osu.Server.Spectator/Hubs/ScoreUploader.cs:117]: Is there a reason why this is preferring the db-retrieved `ScoreInfo` over the already-existing local one? Seems generally arbitrary.

Comment (smoogipoo) [reply osu.Server.Spectator/Hubs/ScoreUploader.cs:117]: The alternative is to manually copy over the `Passed` value from the databased model. Does that sit better with you?

For me, if we have to do that for `Passed` then it seems as something we would have to do over again in the future for other properties, and I feel uneasy about that.

File: osu.Server.Spectator/Database/DatabaseAccess.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Server.Spectator/Database/DatabaseAccess.cs:306]: I'm okay waiting, I don't see this as a higher priority than the other PR.

Comment (bdach) [reply osu.Server.Spectator/Database/DatabaseAccess.cs:306]: disregard above, see https://github.com/ppy/osu-infrastructure/issues/24

**Review by peppy — APPROVED** [review]

**Pull request merged**